gasprice/gasprice.go

```diff
@@ -0,0 +1,265 @@
+//   - collects at most the sample number of sorted transactions in block
+//   - if not enough transactions were collected and their tips, go through some more blocks to get
+//     more accurate calculation
+//   - when enough transactions and their tips are collected, take the one that is in pricePercentile
+//   - if given price is larger then maxPrice then return the maxPrice
+func (g *GasHelper) MaxPriorityFeePerGas() (*big.Int, error) {
+	currentHeader := g.backend.Header()
+
+	currentBlock, found := g.backend.GetBlockByHash(currentHeader.Hash, true)
+	if !found {
+		return nil, fmt.Errorf(couldNotFoundBlockFormat, currentHeader.Number, currentHeader.Hash)
+	}
+
+	g.lock.Lock()
+	lastPrice := g.lastPrice
+	lastHeader := g.lastHeaderHash
+	g.lock.Unlock()
+
+	if currentHeader.Hash == lastHeader {
+		// small optimization, if we calculated already the price for given block
+		return new(big.Int).Set(lastPrice), nil
+	}
+
+	var allPrices []*big.Int
+
+	collectPrices := func(block *types.Block) error {
+		baseFee := block.Header.BaseFee
+		txSorter := newTxByEffectiveTipSorter(block.Transactions, baseFee)
+		sort.Sort(txSorter)
+
+		blockMiner := types.BytesToAddress(block.Header.Miner)
+		signer := crypto.NewSigner(g.backend.Config().Forks.At(block.Number()),
+			uint64(g.backend.Config().ChainID))
+		blockTxPrices := make([]*big.Int, 0)
+
+		for _, tx := range txSorter.txs {
+			tip := tx.EffectiveTip(baseFee)
+
+			if tip.Cmp(g.ignorePrice) == -1 {
+				// ignore transactions with tip lower then ignore price
+				continue
+			}
+
+			sender, err := signer.Sender(tx)
+			if err != nil {
+				return fmt.Errorf("could not get sender of transaction: %s. Error: %w", tx.Hash, err)
+			}
+
+			if sender != blockMiner {
+				blockTxPrices = append(blockTxPrices, tip)
+
+				// if sample number of txs from block is reached,
+				// don't process any more txs
+				if len(blockTxPrices) >= int(g.sampleNumber) {
+					break
+				}
+			}
+		}
+
+		if len(blockTxPrices) == 0 {
+			// either block is empty or all transactions in block are sent by the miner.
+			// in this case add the latests calculated price for sampling
+			blockTxPrices = append(blockTxPrices, lastPrice)
+		}
+
+		// add the block prices to the slice of all prices
+		allPrices = append(allPrices, blockTxPrices...)
+
+		return nil
+	}
+
+	// iterate from current block to previous blocks determined by numOfBlocksToCheck
+	// if chain doesn't have that many blocks, we need to stop the loop (currentBlock.Number() > 0)
+	for i := uint64(0); i < g.numOfBlocksToCheck && currentBlock.Number() > 0; i++ {
+		if err := collectPrices(currentBlock); err != nil {
+			return nil, err
+		}
+
+		currentBlock, found = g.backend.GetBlockByHash(currentBlock.ParentHash(), true)
+		if !found {
+			return nil, fmt.Errorf(couldNotFoundBlockFormat, currentHeader.Number, currentHeader.Hash)
+		}
+	}
+
+	// at least amount of transactions to get
+	minNumOfTx := int(g.numOfBlocksToCheck) * 2
```

Comment:
Why you don't rely on `sampleNumber` parameter here instead?

Reply:
`sampleNumber` is the amount of transactions we pool from the block. This `minNumOfTx` is from `geth` it basically means, if I didn't get at least 2 transactions per block, try to get more transactions from more blocks so that we can get a more precise calculation. If I had empty blocks, or blocks which have only transactions added by miner, then we need to look at more blocks to get enough transactions.

Reply:
Ok, but still I'm thinking about why we would not provide a configuration parameter that would allow us to specify how many transactions per block we would like to gather?

Reply:
You have `sampleNumber`, but as I said, some blocks might be empty, some blocks can have only the miners transactions, so you will not get appropriate amount of transactions. This can also happen if the chain is young (has small amount of blocks), but in the case we do have enough blocks, we would need to look more in history to get more transactions, and this is more then enough.